File: src/spec/resources/testprojects/rubytest/lib/petstore_client/auth/oauth/oauth2_implicit_authenticator.rb
# frozen_string_literal: true

module PetstoreClient
  module Auth
    module OAuth
      # Authenticator for OAuth2 implicit flow.
      class OAuth2ImplicitAuthenticator < Authenticator
        attr_reader :host
        attr_writer :access_token

        def initialize(host, authorization_url, scopes)
          super()
          @host = host
          @authorization_url = authorization_url
          @scopes = scopes.freeze
          @access_token = nil
        end

        def build_authorization_url(state = nil)
          params = { 'response_type' => 'token' }
          params['scope'] = @scopes.join(' ') unless @scopes.empty?
          params['state'] = state if state
          "#{@authorization_url}?#{URI.encode_www_form(params)}"
        end

        def auth_headers
          raise 'Must set access_token before making API requests' if @access_token.nil?

          { 'Authorization' => "Bearer #{@access_token}" }
        end
      end
    end
  end
end

File: src/spec/resources/testprojects/nodetest/src/auth/oauth/oauth2-implicit-authenticator.ts
import { BaseAuthenticator } from '../base-authenticator.js';

/**
 * Authenticator for the OAuth2 Implicit flow.
 */
export class OAuth2ImplicitAuthenticator extends BaseAuthenticator {
  private readonly host: string;
  private readonly authorizationUrl: string;
  private readonly scopes: readonly string[];
  private accessToken: string | null = null;

  constructor(host: string, authorizationUrl: string, scopes: string[]) {
    super();
    this.host = host;
    this.authorizationUrl = authorizationUrl;
    this.scopes = Object.freeze([...scopes]);
  }

  buildAuthorizationUrl(state?: string): string {
    const params = new URLSearchParams({ response_type: 'token' });
    if (this.scopes.length > 0) {
      params.set('scope', this.scopes.join(' '));
    }
    if (state) {
      params.set('state', state);
    }
    return `${this.authorizationUrl}?${params.toString()}`;
  }

  setAccessToken(token: string): void {
    this.accessToken = token;
  }

  getHost(): string {
    return this.host;
  }

  getAuthHeaders(): Record<string, string> {
    if (!this.accessToken) {
      throw new Error('Must call setAccessToken() before making API requests');
    }
    return { Authorization: `Bearer ${this.accessToken}` };
  }
}
